### PR TITLE
Apply blackdoc to python source

### PR DESCRIPTION
**Pull request by akaszynski**
PyVista has adopted [black](https://github.com/psf/black), but we fail to implement this for our documentation or doc strings.

This PR implements [blackdoc](https://github.com/keewis/blackdoc) to reformat all Python code in our docstrings as a `pre-commit`.

Pinging @jorgepiloto because this is really cool and IMO works better than [blacken-docs](https://pypi.org/project/blacken-docs/)

### CODE REVIEW

**Review by tkoyama010 — APPROVED** [review]
Excellent. I was always concerned that the docstring code was not formatted by black. This solves that!

**Comment by adeak** [comment]
Any chance that we can separate the black run from any other change (i.e. adding this to pre-commit) in separate PRs? I'd like to be able to skip the style PR's merge commit in the git blame. As usual.

**Comment by adeak** [comment]
To be clear the ideal situation would be to separate all autoformatting. If you changed anything manually and know what that was, that should also be moved along with the other changes. With one big commit that's probably not feasible though.

And along with the pre-commit addition we should add a check for this in CI, otherwise noncomforming commits will startvseeping back quick.

**Comment by tkoyama010** [comment]
> And along with the pre-commit addition we should add a check for this in CI, otherwise noncomforming commits will startvseeping back quick.

I guess the pre-commit error will stop CI, isn't it?
https://github.com/pyvista/pyvista/blob/5ff97bbf64f1a4f1f39627db3ccbb7f6d9e71c4b/.github/workflows/style_docstr.yml#L16-L26

**Comment by adeak** [comment]
Aaah, right! Thanks @tkoyama010 👍

**Comment by MatthewFlamm** [comment]
This would be a good example where a merge commit would be better to preserve the commit history in a PR (one commit for each step here), but we have disabled this (for good reasons).

Another thought is that you could have two PRs, one for adding to our pre-commit (this will fail), the other for the changes.  You would merge the first if the second passes and then merge them in one after the other.

A lot of extra work however.

**Comment by adeak** [comment]
> Another thought is that you could have two PRs, one for adding to our pre-commit (this will fail), the other for the changes. You would merge the first if the second passes and then merge them in one after the other.

That's what I had in mind. First merge the autoformatting changes, then adding the pre-commit hook and any hypothetical manual changes needed.

> A lot of extra work however.

Only if @akaszynski touched formatting beyond running blackdoc. The non-autoformat changes would be quite obvious otherwise. And we can only do this right once, now, or we'll be stuck with crap history.

It's of course possible that I'm the only one looking at git blames, in which case "sucks to be me" is another option :P

**Comment by akaszynski** [comment]
Agree with separating this out. I didn't consider git blame, but when I do need it it's quite valuable for tracking down bugs.

**Comment by adeak** [comment]
You could keep most of the changes here, and just revert the few other changes. Whichever is less of a hassle.

**Comment by akaszynski** [comment]
Ok @adeak. I went ahead and opened #4120. I would like to merge the changes from branch into main with and also add `git blame --ignore-rev`.

After the merge I will add the commit hash to the `. git-blame-ignore-revs` file.

See
https://akrabat.com/ignoring-revisions-with-git-blame/
https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/

**Comment by akaszynski** [comment]
Please note that the changes here are entirely from the auto-formatter. Recommending merging this PR first and adding the ignore rev file to #4120.

**Comment by adeak** [comment]
> Please note that the changes here are entirely from the auto-formatter. Recommending merging this PR first and adding the ignore rev file to #4120.

I'm on mobile for the next few hours still, but if this only has autoformatting only then this is good to go. Any manual corrections would have to be put elsewhere anyway.

Adding the ignore-revs item in the follow-up is fine too. For the original blackening we had https://github.com/pyvista/pyvista/pull/2684.

Thanks a lot @akaszynski.

**Review by adeak — APPROVED** [review]

**Comment by MatthewFlamm** [comment]
This sounds like a good plan

**Review by MatthewFlamm — APPROVED** [review]

File: pyvista/core/filters/data_set.py
Comment (adeak) [inline pyvista/core/filters/data_set.py:396]: I wonder how the line length limit exactly works. From [the docs](https://blackdoc.readthedocs.io/en/latest/options.html#black) I assume it uses the same line length as `black`. This doesn't sound ideal though, because I we explicitly strive to have shorter lines for docs [as per CONTRIBUTING.md](https://github.com/pyvista/pyvista/blob/main/CONTRIBUTING.rst#coding-style):

> As for docstrings, follow the guidelines specified in [PEP 8 Maximum Line Length](https://www.python.org/dev/peps/pep-0008/#maximum-line-length) of limiting docstrings to 72 characters per line.

Looking at the blackdoc docs it doesn't seem obvious that we can choose a different line length for blackdoc, but [based on the source](https://github.com/keewis/blackdoc/blob/6fa571ad4c905745bf9574bf08248c60529ebe81/blackdoc/blackcompat.py#L106-L109) it seems that the `[blackdoc]` section in `pyproject.toml` takes precedence (as would be logical):
```py
    black_config = pyproject_toml.get("tool", {}).get("black", {})
    blackdoc_config = pyproject_toml.get("tool", {}).get("blackdoc", {})
    config = {**black_config, **blackdoc_config}
```

So do we want to enforce the 72-char length we mention for docstrings?

Comment (adeak) [reply pyvista/core/filters/data_set.py:396]: Ha, too late :D

File: pyvista/utilities/reader.py
Comment (adeak) [inline pyvista/utilities/reader.py:159]: This is not a blackdoc issue, only a pointer to something potentially redundant: in `pyproject.toml` we have `doctest_optionflags = 'NUMBER ELLIPSIS'` so I wonder if the four `+ELLIPSIS` directives we have in `reader.py` are reduntant. To be looked at later if anyone feels like it.

File: pyvista/utilities/transformations.py
Comment (adeak) [inline pyvista/utilities/transformations.py:250]: I looked through the 45 changed files and this was the only terrible change. I suspect this is mostly a bug, and has to do with the trailing comma. I'd consider removing this trailing comma in the follow-up PR.

File: pyvista/core/filters/data_set.py
Comment (adeak) [reply pyvista/core/filters/data_set.py:396]: Pinging @akaszynski just to make sure we discuss this (and the other, far less important points) before going forward.

**Comment by adeak** [comment]
Also pointing out for posterity that the changes here were made with an explicit exclusion of `pyvista/plotting/charts.py`, see https://github.com/pyvista/pyvista/pull/4120#discussion_r1133144610.

**Comment by akaszynski** [comment]
I’ll work on the follow up PR and implement the changes before enforcing blackdox. Thanks for the comments, I’ll be less trigger happy…

**Comment by adeak** [comment]
> I’ll be less trigger happy…

It's all good, I wouldn't have approved if I had insisted on taking a close look before merge :) In the worst case of changing the config and rerunning `blackdoc` on the whole codebase we can just have another autoformat PR and add _two_ commits to the ignore-revs file. No harm done.

File: pyvista/core/filters/data_set.py
Comment (adeak) [reply pyvista/core/filters/data_set.py:396]: Hmm, yeah, too bad [the problem seems to be a hornet's nest](https://stackoverflow.com/a/71041192/5067311).

File: pyvista/utilities/transformations.py
Comment (akaszynski) [reply pyvista/utilities/transformations.py:250]: Resolved in https://github.com/pyvista/pyvista/pull/4129.

File: pyvista/core/filters/data_set.py
Comment (akaszynski) [reply pyvista/core/filters/data_set.py:396]: Discussed in https://github.com/pyvista/pyvista/pull/4129

**Pull request merged**
